### ISSUE TITLE
Яндекс.Практикум:Код-ревью

### ISSUE THREAD

**Issue by zlocate** [rsungatovich/rsungatovich.github.io#4]
Добрый вечер. Благодарю за проделанную работу.
К сожалению, текущая реализация содержит ряд недостатков не позволяющих мне засчитать ее на данной итерации:

## Репозиторий

**Нужно исправить** файл README должен содержать активную (кликабельную) ссылку на проект и короткое описание, раскрывающий суть работы \ назначение проекта.

**Нужно исправить**: отсутствуют git-теги отражающие номер версии. Создать их можно при помощи команды `git tag -a v0.0.1 -m "my first version"`
Обращаю внимание, что по умолчанию теги не публикуются при вызове команды `git push`
Для их публикации выполните команду `git push origin --tags`
Подробнее: https://git-scm.com/book/ru/v2/%D0%9E%D1%81%D0%BD%D0%BE%D0%B2%D1%8B-Git-%D0%A0%D0%B0%D0%B1%D0%BE%D1%82%D0%B0-%D1%81-%D0%BC%D0%B5%D1%82%D0%BA%D0%B0%D0%BC%D0%B8

**Можно лучше**: Добавить .gitignore.
.gitignore должен содержать правила, исключающие попадание временных и\или системных файлов в репозиторий.
В случае если вы затрудняетесь с составлением списка базовых правил можете воспользоваться сервисом-генератором, например gitignore.io

## Регулярные выражения
**Можно лучше** часть регулярного выражения, описывающую протокол в адресе сайта можно оформить более лаконично - `https?`
**Можно лучше** Квантификатор `{1}` можно не использовать. Он не играет никакой роли.
**Можно лучше** Квантификатор {0,} можно заменить на *
Подробнее о квантификаторах: https://learn.javascript.ru/regexp-quantifiers

### Email
**Нужно исправить** следующие некорректные кейсы проходят валидацию
`_@site.ru` - имя ящика не может состоять только из спецсимволов
`_some_@ya.ru` - имя ящика не может начинаться и заканчиваться на спецсимвол

### Номер телефона
**Нужно исправить**: Данные форматы номеров телефонов проходят валидацию, хотя с точки зрения постановки задачи они являются невалидными.
```
+7925900-90-90
+7 (925) 900-90-90
+7 (925) 9009090
+7(925) 9009090
+7 (925)9009090
+7(925)9009090
+7 925 900-90-90
```

Адрес сайта
**Нужно исправить** следующие некорректные кейсы проходят валидацию
`http://127.0..`  - некорректный IP, точка на конце
`http://127.0.0..` - некорректный IP, точка на конце

**Можно лучше** следующие некорректные кейсы проходят валидацию
`http://www.google.ru///////` (в пути количество слешей подряд должно быть ограничено 1)

**Comment by rsungatovich** [comment]
Добрый день. Спасибо за ревью!

Все критические замечания были исправлены. 

- Хотел бы отметить один кейс. В моих паттернах в номерах телефона значения +7 и 8 это одно и тоже, поэтому они могут встречаться в общих конструкциях, что на мой взгляд логично.
(по поводу этого замечаний небыло в предыдущем ревью, но все же внесу ясности на всякий случай)

Все пожелания "можно лучше" были учтены, кроме:

**1.** **Можно лучше** часть регулярного выражения, описывающую протокол в адресе сайта можно оформить более лаконично - https?

- В таком случае протокол будет не обязательным, а в задании он необходим. Возможно вы имели ввиду сделать как-то иначе?

**2.** **Можно лучше** следующие некорректные кейсы проходят валидацию
http://www.google.ru/////// (в пути количество слешей подряд должно быть ограничено 1)

- К сожалению, я не нашел нормального способа, как это можно улучшить. Не стал отправлять результат, который мне не нравится, может быть вы могли бы посоветовать как это лучше реализовать?

**Comment by zlocate** [comment]
> Добрый день. Спасибо за ревью!
> 
> Все критические замечания были исправлены.
> 
> * Хотел бы отметить один кейс. В моих паттернах в номерах телефона значения +7 и 8 это одно и тоже, поэтому они могут встречаться в общих конструкциях, что на мой взгляд логично.
>   (по поводу этого замечаний небыло в предыдущем ревью, но все же внесу ясности на всякий случай)
> 
> Все пожелания "можно лучше" были учтены, кроме:
> 
> **1.** **Можно лучше** часть регулярного выражения, описывающую протокол в адресе сайта можно оформить более лаконично - https?
> 
> * В таком случае протокол будет не обязательным, а в задании он необходим. Возможно вы имели ввиду сделать как-то иначе?
> 
> **2.** **Можно лучше** следующие некорректные кейсы проходят валидацию
> http://www.google.ru/////// (в пути количество слешей подряд должно быть ограничено 1)
> 
> * К сожалению, я не нашел нормального способа, как это можно улучшить. Не стал отправлять результат, который мне не нравится, может быть вы могли бы посоветовать как это лучше реализовать?

Добрый день. Благодарю за оперативное внесение изменений.

1. Нет, исходя из правил написания регулярного выражения символ ? накладывает ограничение на последний символ или оформленный блок. Если бы мы оформили https в скобки, тогда да - это бы означало что протокол может как быть так и отсутствовать.
Мой же вариант описывает вариант когда допускается наличие или отсутствие единственной буквы `s`.
Это не совсем очевидно :)

2. Можно отталкиваться от чего-то подобного как от идеи
`((\/?[a-z0-9]+\/?)*|(?!\/)\/)#?`

**Comment by zlocate** [comment]
По номеру телефона, к сожалению, вынужден вас попросить жестко зафиксировать вариант `+7`.
Ваша мотивация мне ясна, однако в текущей формулировке задания указаны маски номеров, которые должны проходить валидацию. Все отличные от них считаются некорректными.
Еще про популярный вопрос от студентов - ремарка про пробелы может сбивать с толку, однако ее суть обратить внимание на то, что в масках представленных в задании они (пробелы) есть.
Мы, также как и вы, работаем в рамках ограничений накладываемы заданием, однако мы работает над тем чтобы они становились лучше.
Благодарю за понимание

**Comment by rsungatovich** [comment]
**Нужно исправить** следующие корректные адреса сайта не проходят валидацию
http://some.test.yandex.ru , https://some.amazing.site.name.academy.some - адрес сайта может содержать произвольное число поддоменов.

Они не проходят потому что в ТЗ указаны вот такие примеры. Вы же сами пишете что нужно четко соответствовать ТЗ. Тогда почему в этом задании необходимо отклонятся от примеров в ТЗ, а в задании с номерами телефонов нет?

http://ya.ru
https://www.ya.ru
http://2-domains.ru
http://ya.ru:98/
http://ya.com:30
http://ya.ru/path/to/deep/
http://ya-ya-ya.ru
http://8.8.8.8:8080
http://8.8.8.8:8080/page/to/deep#

**Comment by rsungatovich** [comment]
Были исправлены все ваши замечания, но вопрос выше все также актуален. 

Спасибо за ревью!

**Comment by zlocate** [comment]
> **Нужно исправить** следующие корректные адреса сайта не проходят валидацию
> http://some.test.yandex.ru , https://some.amazing.site.name.academy.some - адрес сайта может содержать произвольное число поддоменов.
> 
> Они не проходят потому что в ТЗ указаны вот такие примеры. Вы же сами пишете что нужно четко соответствовать ТЗ. Тогда почему в этом задании необходимо отклонятся от примеров в ТЗ, а в задании с номерами телефонов нет?
> 
> http://ya.ru
> https://www.ya.ru
> http://2-domains.ru
> http://ya.ru:98/
> http://ya.com:30
> http://ya.ru/path/to/deep/
> http://ya-ya-ya.ru
> http://8.8.8.8:8080
> http://8.8.8.8:8080/page/to/deep#

Это проблема текущей формулировки задания, исправляем .  (однако ревьюверы влияют на сам текст обособлено исходя из вашего фидбека и не могут быстро вносить правки напрямую)

Разъясню нашу позицию.
Для номеров телефонов, в отличие от адреса сайта возможно указать полный список возможных вариантов адекватной длины.
Вариативность для каждой из компонент достаточна чтобы общее число возможных вариантов было крайне велико.

В случае тонких и спорных либо редких вариантов я готов снизить критичность, однако в данном случае это крайне частый с практической точки зрения кейс. 
Данный кейс важен еще, и по тому, что потенциальный паблишер формы с значительной долей вероятности просто не сможет оставить заявку ибо она не пройдет валидацию из-за сайта с поддоменом.
Согласитесь, не самый приятный сценарий.

На ваш взгляд, адрес сайта http://mail.yandex.ru является корректным и должен проходить проверку или нет?

**Comment by zlocate** [comment]
Работя принята. Желаю успехов в дальнейшем изучении курса.

**Comment by rsungatovich** [comment]
Я согласен с вашими замечаниями, что ссылки должны иметь поддомены. Шаблон, который учитывает поддомены у меня был еще в самом начале, но после того, когда вы написали про то, что работа должна быть выполнена четко по указанным примерам из ТЗ, я изменил этот шаблон, убрав с него поддомены. 

Вопрос как раз таки в противоречии того, что здесь мы учитываем определенные кейсы, а в других ситуациях мы их не используем. Без четкого ТЗ, данное решение остается субъективным, тем более для человека не имеющего большего опыта, чтобы принимать подобные решения самостоятельного.

Спасибо!
